### PR TITLE
Learning system in the REPL 3

### PR DESCRIPTION
**Pull request by teodorlu**
Why 3? Because 1 and 2 got merged into master.

High level ideas:

1. Collect and curate exercises relevant to Athens
2. Track user progress -- what exercises are completed, and not?
3. Let the user customize their learning path; choose track, skip tasks, etc.

### CODE REVIEW

**Comment by teodorlu** [comment]
Just pushed a first attempt to use Datascript as an in-memory task store. For now, that means we can write datascript queries to get tasks out.

Example usage:

```clojure
(ns clojurefam.sandbox.teodorlu.tasks
  (:require [clojurefam.task :as task]
            [clojurefam.db :as db]))

;; Source: http://www.4clojure.com/problem/9
(task/def ::conj-1
  {:task/problem '(= #{1 2 3 4} (conj #{1 4 3} __))
   :task/check (fn [solution]
                 (= #{1 2 3 4} (conj #{1 4 3} solution)))})

;; Source: http://www.4clojure.com/problem/10
(task/def ::maps-1
  {:task/problem '(and
                   (= __ ((hash-map :a 10, :b 20, :c 30) :b))
                   (= __ (:b {:a 10, :b 20, :c 30})))
   :task/check (fn [solution]
                 (and
                  (= solution ((hash-map :a 10, :b 20, :c 30) :b))
                  (= solution (:b {:a 10, :b 20, :c 30}))))})

(comment
  (db/pull '[*] [:task/id ::conj-1])
  ;; => {:db/id 1,
  ;;     :task/check #function[clojurefam.sandbox.teodorlu.tasks/eval17861/fn--17862],
  ;;     :task/id :clojurefam.sandbox.teodorlu.tasks/conj-1,
  ;;     :task/problem (= #{1 4 3 2} (conj #{1 4 3} __))}

  (db/pull '[:task/problem] [:task/id ::conj-1])
  ;; => #:task{:problem (= #{1 4 3 2} (conj #{1 4 3} __))}
  )
```

**Comment by teodorlu** [comment]
I've also ditched the longer `athens.scratch.x` for `clojurefam.x`.

**Comment by teodorlu** [comment]
Refactored: single `koan` namespace.

- `koan/defkoan`: define a Clojure koan
- `koan/help`: show information about the koan
- `koan/defsolution`: solve a koan

Example usage: https://github.com/athensresearch/ClojureFam/blob/44dc68da0b3a84ec4105f00f49447e637f0739a7/src/clojurefam/sandbox/teodorlu/solutions.clj#L15

**Comment by teodorlu** [comment]
I'm considering if we should remove `koan/defsolution` and just provide `koan/check` instead.

**Comment by Bardia95** [comment]
@teodorlu everything looks great! I think we should remove `koan/defsolution` and just provide `koan/check` instead, like you suggested. But what was the original reasoning for including `koan/defsolution`?

File: src/athens/sandbox/api.clj [hunk 1]
```diff
@@ -1,25 +1,94 @@
 (ns ^{:doc "End-user API expected to be exposed to learners in the REPL"}
```
Comment (Bardia95) [inline src/athens/sandbox/api.clj:1]: Are we scrapping everything in the Athens sandbox and just going with the ClojureFam route? If so, may make sense to just comment or remove any files in here.

**Comment by teodorlu** [comment]
On macro (`defsolution`) vs function (`check`), I think both are acceptable.

- Check is simpler, "start narrow, expand later".
- If we keep `defsolution` as a macro, we can implement registration later without students seeing anything.

However, `check` is probably the least surprising.

**Comment by teodorlu** [comment]
> If we are going with `koan` instead of `task` as our terminology, do you think it makes sense to change the name of the `/clojurefam/sandbox/teodorlu/tasks.clj` file to `/clojurefam/sandbox/teodorlu/koans.clj`?

Valid point.

- `defkoan` implies the ns should be `[...].koans`
- However, is a koan just the problem, or is the koan also the solution? I guess we could just use `[...].solutions`.

In conclusion, I agree: we should rename.

File: src/athens/sandbox/api.clj (hunk 1 above)
Comment (teodorlu) [reply src/athens/sandbox/api.clj:1]: Hmm, I agree.

I think we're going to need the part about datascript DB loading later when writing Roam db koans. But we can easily put that back at a later point.

Yup, let's remove it.

**Comment by teodorlu** [comment]
@Bardia95, please go ahead if you'd like to fix the points we've discussed above (just push to this branch, I have no unmerged work locally). Otherwise, I'll get to it after work. Let me know if you start, so that we can avoid double work..

**Comment by Bardia95** [comment]
Just got to this now @teodorlu, I already made another branch on top of this one, so I will finish up with the exercises I'm adding and then I'll merge this all back into master

**Comment by teodorlu** [comment]
> Just got to this now @teodorlu, I already made another branch on top of this one, so I will finish up with the exercises I'm adding and then I'll merge this all back into master

@Bardia95 if you're planning to do the merge (instead of pushing to this branch), what is your opinion on `check` vs `defsolution`?

> On macro (`defsolution`) vs function (`check`), I think both are acceptable.
> 
>     * Check is simpler, "start narrow, expand later".
> 
>     * If we keep `defsolution` as a macro, we can implement registration later without students seeing anything.
> 
> 
> However, `check` is probably the least surprising.

**Comment by Bardia95** [comment]
Sorry forgot to address that—let’s just keep in defsolution for now, no harm in it really

**Comment by teodorlu** [comment]
OK. Merge at your leisure!

**Comment by teodorlu** [comment]
OK. Your call when to merge!

**Pull request merged**